### PR TITLE
feat: Add implementation for Orange & Rockland Utilities

### PR DESCRIPTION
**Pull request by bvlaicu**
[Orange & Rockland Utilities](oru.com) (ORU) provides electric and gas service for more than 300,000 households and businesses in six counties in New York and northern New Jersey.

ORU is owned by ConEd so their implementations are pretty similar.

Example output using this new utility:

image

### CODE REVIEW

**Review by tronikos — CHANGES_REQUESTED** [review]

File: src/opower/utilities/oru.py
[the remark this answers is not in the document]
Comment (bvlaicu) [reply src/opower/utilities/oru.py:1]: Makes sense. Done. Thanks.

File: src/opower/utilities/oru.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+RETURN_URL = "/en/accounts-billing/my-account/energy-use"
+
+
+class Oru(UtilityBase):
+    """Orange & Rockland Utilities (Oru)."""
+
+    @staticmethod
+    def name() -> str:
+        """Distinct recognizable name of the utility."""
+        return "Orange & Rockland Utilities (Oru)"
```
[the remark this answers is not in the document]
Comment (bvlaicu) [reply src/opower/utilities/oru.py]: Correct. Done.

File: src/opower/utilities/oru.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+import aiohttp
+from pyotp import TOTP
+
+from ..const import USER_AGENT
+from ..exceptions import InvalidAuth
+from .base import UtilityBase
+from .coned import ConEd
+
+HOSTNAME = "oru.com"
```
Comment (tronikos) [inline src/opower/utilities/oru.py:13]: I don't see how this overrides the one in coned.py. Have you tested your latest revision still works? If it does then I guess you can login to coned.com with oru.com credentials in which case you don't need any changes in coned.py and just go with your changes below. If it doesn't work, you have to extract the hostname to a static method and override it below.

File: src/opower/utilities/oru.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+class Oru(ConEd, UtilityBase):
```
Comment (tronikos) [inline src/opower/utilities/oru.py:15]: You shouldn't need to extend UtilityBase since ConEd already extends it.

**Review by tronikos — CHANGES_REQUESTED** [review]

File: src/opower/utilities/oru.py (hunk 2 above)
Comment (tronikos) [reply src/opower/utilities/oru.py:13]: You will have to change only in coned.py:

from

```python
    @staticmethod
    async def async_login(
        session: aiohttp.ClientSession,
```
to:

```python
    @classmethod
    async def async_login(
        cls,
        session: aiohttp.ClientSession,
```

and then inside async_login you will be able to reference the @staticmethod that returns the hostname with: `cls.hostname()`

This also means some of the constants in coned.py will have to become variables in async_login.

**Comment by tronikos** [comment]
I committed https://github.com/tronikos/opower/commit/a580b15953b5355ef41ebc2f507a53ffc1ffee2c with your changes plus my suggestions. Thanks!

**Comment by bvlaicu** [comment]
Oh, thanks, I just committed my changes then saw your updates. They are very similar, anyway.
And regarding testing my previous version, yes, I tested it so yes, you can log into coned with oru creds. Which I kind of knew from other past experiments if I think about it now.
Thanks for all your work.

**Pull request closed without merging**